lib/Scheduler/aco.cpp

```diff
@@ -251,14 +253,59 @@ FUNC_RESULT ACOScheduler::FindSchedule(InstSchedule *schedule_out,
                                        SchedRegion *region) {
   rgn_ = region;
 
+  if(!aco_use_two_pass)
+  {
+    pass = ACOPass::ALL_IN_ONE;
+    return performSchedPass(schedule_out);
+  }
+  else
+  {
+    //in the first pass we schedule for register pressure only
+    //in the second pass we schedule for lower schedule length
+    //but only accept schedules with an RP <= the first passes RP
+    //the first pass result is the initial schedule the second pass
+    pass = ACOPass::FIRST;
+    FUNC_RESULT pass1Res = performSchedPass(schedule_out);
+	if(pass1Res!=RES_SUCCESS) {
+      return pass1Res;
+    }
+    pass = ACOPass::SECOND;
+    setInitialSched(schedule_out);
+    return performSchedPass(schedule_out);
+
+  }
+}
+
+//compares to schedules taking into account which pass
+//we are on and returns whether the old schedule
+//is inferior to the new one and ought to be replaced
+bool ACOScheduler::shouldReplaceSchedule(ACOPass pass, InstSchedule* oldSched,
+                                         InstSchedule* newSched) {
+  //Logger::Info("old_sched_addr:%p, new_sched_addr:%p", oldSched, newSched);
+  return newSched!=NULL &&
+         (oldSched==NULL ||
+         (pass==ACOPass::FIRST &&
+         newSched->GetSpillCost() < oldSched->GetSpillCost()) ||
+         (pass==ACOPass::SECOND && newSched->GetCost() < oldSched->GetCost()
+         && newSched->GetSpillCost() <= oldSched->GetSpillCost()) ||
+         (pass==ACOPass::ALL_IN_ONE &&
+         newSched->GetCost() < oldSched->GetCost()));
+
+}
+
+FUNC_RESULT ACOScheduler::performSchedPass(InstSchedule *schedule_out) {
+
   // initialize pheremone
   // for this, we need the cost of the pure heuristic schedule
   int pheremone_size = (count_ + 1) * count_;
   for (int i = 0; i < pheremone_size; i++)
     pheremone_[i] = 1;
   initialValue_ = 1;
+  InstSchedule* heuSched = FindOneSchedule();
```

Comment:
Unique ptr

Reply:
ACOScheduler::FindOneSchedule() returns a raw pointer to a new allocated InstScheule.  Previously this was leaked.  Code before this commit would call GetCost() on the schedule returned without deleting it.  I agree that we should use unique_ptr s in the future, but the scope of this commit was adding the code for two pass ACO.  I intend to move ACOScheduler over to using unique_ptr s in the future.